packages/gatsby-source-contentful/src/on-pre-bootstrap.js

```diff
@@ -0,0 +1,352 @@
+// @todo import syntax!
+import normalize from "./normalize"
+const isOnline = require(`is-online`)
+const _ = require(`lodash`)
+const fs = require(`fs-extra`)
+const { createClient } = require(`contentful`)
+const v8 = require(`v8`)
+const { CODES } = require(`./report`)
+const fetchData = require(`./fetch`)
+const { createPluginConfig } = require(`./plugin-options`)
+
+const restrictedContentTypes = [`entity`, `reference`]
+
+export async function onPreBootstrap(
+  { reporter, cache, actions, parentSpan, getNode, getNodes, createNodeId },
+  pluginOptions
+) {
+  // Fetch data for sourceNodes & createSchemaCustomization
+  const { deleteNode, touchNode } = actions
+
+  let currentSyncData
+  let contentTypeItems
+  let defaultLocale
+  let locales
+  let space
+  if (process.env.GATSBY_CONTENTFUL_EXPERIMENTAL_FORCE_CACHE) {
+    reporter.info(
+      `GATSBY_CONTENTFUL_EXPERIMENTAL_FORCE_CACHE: Storing/loading remote data through \`` +
+        process.env.GATSBY_CONTENTFUL_EXPERIMENTAL_FORCE_CACHE +
+        `\` so Remote changes CAN NOT be detected!`
+    )
+  }
+  const forceCache = await fs.exists(
+    process.env.GATSBY_CONTENTFUL_EXPERIMENTAL_FORCE_CACHE
+  )
+
+  const pluginConfig = createPluginConfig(pluginOptions)
+  const sourceId = `${pluginConfig.get(`spaceId`)}-${pluginConfig.get(
+    `environment`
+  )}`
+
+  const CACHE_SYNC_TOKEN = `contentful-sync-token-${sourceId}`
+  const CACHE_SYNC_DATA = `contentful-sync-data-${sourceId}`
+
+  /*
+   * Subsequent calls of Contentfuls sync API return only changed data.
+   *
+   * In some cases, especially when using rich-text fields, there can be data
+   * missing from referenced entries. This breaks the reference matching.
+   *
+   * To workround this, we cache the initial sync data and merge it
+   * with all data from subsequent syncs. Afterwards the references get
+   * resolved via the Contentful JS SDK.
+   */
+  const syncToken = await cache.get(CACHE_SYNC_TOKEN)
+  let previousSyncData = {
+    assets: [],
+    entries: [],
+  }
+  const cachedData = await cache.get(CACHE_SYNC_DATA)
+
+  if (cachedData) {
+    previousSyncData = cachedData
+  }
+
+  const fetchActivity = reporter.activityTimer(
+    `Contentful: Fetch data (${sourceId})`,
+    {
+      parentSpan,
+    }
+  )
+
+  // If the cache has data, use it. Otherwise do a remote fetch anyways and prime the cache now.
+  // If present, do NOT contact contentful, skip the round trips entirely
+  if (forceCache) {
```

Comment:
This experimental flag adds some complexity to this code we could avoid.

I can't find any documentation to it as well